// tailwindcss-3d/css-animations/spin.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Spin", {
    enumerable: true,
    get: function() {
        return Spin;
    }
});
var _lodash = require("lodash");
var _every = /*#__PURE__*/ _interop_require_default(require("lodash/every"));
var _isPlainObject = /*#__PURE__*/ _interop_require_default(require("lodash/isPlainObject"));
var _isString = /*#__PURE__*/ _interop_require_default(require("lodash/isString"));
var _isUndefined = /*#__PURE__*/ _interop_require_default(require("lodash/isUndefined"));
var _keys = /*#__PURE__*/ _interop_require_default(require("lodash/keys"));
var _mapValues = /*#__PURE__*/ _interop_require_default(require("lodash/mapValues"));
var _values = /*#__PURE__*/ _interop_require_default(require("lodash/values"));
var _defaultTheme = /*#__PURE__*/ _interop_require_default(require("tailwindcss/defaultTheme"));
var _base = require("./base");
var _transform = require("../css-utilities/transform");
var _cssvalue = require("../utils/css-value");
var _generateguard = require("../utils/generate-guard");
var _lodashtransformers = require("../utils/lodash-transformers");
function _assert_this_initialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _get_prototype_of(o) {
    _get_prototype_of = Object.setPrototypeOf ? Object.getPrototypeOf : function getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _get_prototype_of(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _set_prototype_of(subClass, superClass);
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function _possible_constructor_return(self, call) {
    if (call && (_type_of(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assert_this_initialized(self);
}
function _set_prototype_of(o, p) {
    _set_prototype_of = Object.setPrototypeOf || function setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _set_prototype_of(o, p);
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _is_native_reflect_construct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;
    try {
        Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
        return true;
    } catch (e) {
        return false;
    }
}
function _create_super(Derived) {
    var hasNativeReflectConstruct = _is_native_reflect_construct();
    return function _createSuperInternal() {
        var Super = _get_prototype_of(Derived), result;
        if (hasNativeReflectConstruct) {
            var NewTarget = _get_prototype_of(this).constructor;
            result = Reflect.construct(Super, arguments, NewTarget);
        } else {
            result = Super.apply(this, arguments);
        }
        return _possible_constructor_return(this, result);
    };
}
var Spin = /*#__PURE__*/ function(Base) {
    "use strict";
    _inherits(Spin, Base);
    var _super = _create_super(Spin);
    function Spin() {
        _class_call_check(this, Spin);
        var _this;
        _this = _super.apply(this, arguments);
        _define_property(_assert_this_initialized(_this), "isProcessableValue", (0, _generateguard.generateGuard)(_isString.default, _isUndefined.default));
        _define_property(_assert_this_initialized(_this), "isProcessableValues", (0, _generateguard.generateGuard)([
            _isPlainObject.default,
            function(maybe) {
                return (0, _every.default)((0, _keys.default)(maybe), _isString.default);
            },
            function(maybe) {
                return (0, _every.default)((0, _values.default)(maybe), _this.isProcessableValue);
            }
        ]));
        _define_property(_assert_this_initialized(_this), "normaliseValues", function(values) {
            return _this.isProcessableValues(values) ? (0, _lodash.chain)(values).mapKeys(function(_duration, modifier) {
                return (0, _cssvalue.normaliseNumberValue)(modifier);
            }).mapValues(function(duration) {
                return (0, _cssvalue.normaliseTimeValue)(duration);
            }).pickBy(function(duration, modifier) {
                return (0, _isString.default)(modifier) && modifier !== "" && (0, _isString.default)(duration);
            }).value() : {};
        });
        _define_property(_assert_this_initialized(_this), "defaultTheme", _this.normaliseValues((0, _mapValues.default)(_defaultTheme.default.spacing, function(_spacing, modifier) {
            return "".concat(modifier, "s");
        })));
        _define_property(_assert_this_initialized(_this), "xyKeyframes", function() {
            var rotateDeclarations = function(axis, value) {
                var rotateProperty = "rotate".concat(axis.toUpperCase());
                var transformValue = _this.legacy ? _transform.Transform.normaliseLegacyFunctionValues(_define_property({}, rotateProperty, value)) : _transform.Transform.normaliseFunctionValues(_define_property({}, rotateProperty, value));
                return {
                    "--webkit-transform": transformValue,
                    transform: transformValue
                };
            };
            return (0, _lodash.chain)({
                "1": ""
            }).transform((0, _lodashtransformers.axesModifier)([
                "x",
                "y"
            ]), {}).transform((0, _lodashtransformers.nameModifier)("spin"), {}).transform((0, _lodashtransformers.signModifier)(), {}).mapValues(function(param) {
                var axis = param.axis, sign = param.sign;
                return {
                    from: _object_spread({}, rotateDeclarations(axis, "0deg")),
                    to: _object_spread({}, rotateDeclarations(axis, "".concat(sign, "360deg")))
                };
            }).value();
        });
        _define_property(_assert_this_initialized(_this), "zKeyframes", function() {
            var rotateDeclarations = function(value) {
                if (!_this.legacy) {
                    return {
                        rotate: value
                    };
                }
                var transformValue = _transform.Transform.normaliseLegacyFunctionValues({
                    rotateZ: value
                });
                return {
                    "--webkit-transform": transformValue,
                    transform: transformValue
                };
            };
            return (0, _lodash.chain)({
                "1": ""
            }).transform((0, _lodashtransformers.axesModifier)("z"), {}).transform((0, _lodashtransformers.nameModifier)("spin"), {}).transform((0, _lodashtransformers.signModifier)(), {}).mapValues(function(param) {
                var sign = param.sign;
                return {
                    from: _object_spread({}, rotateDeclarations("0deg")),
                    to: _object_spread({}, rotateDeclarations("".concat(sign, "360deg")))
                };
            }).value();
        });
        _define_property(_assert_this_initialized(_this), "keyframes", function() {
            return _object_spread({}, _this.xyKeyframes(), _this.zKeyframes());
        });
        _define_property(_assert_this_initialized(_this), "animation", function(param) {
            var theme = param.theme;
            var values = _this.normaliseValues(theme("spin"));
            return (0, _lodash.chain)(values).transform((0, _lodashtransformers.axesModifier)(), {}).transform((0, _lodashtransformers.nameModifier)("spin"), {}).transform((0, _lodashtransformers.signModifier)(), {}).mapValues(function(param) {
                var value = param.value, axis = param.axis, name = param.name, sign = param.sign;
                return "".concat(sign).concat(name, "-").concat(axis, " ").concat(value, " linear infinite");
            }).value();
        });
        return _this;
    }
    return Spin;
}(_base.Base);
