// tailwindcss-3d/css-utilities/translate.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Translate", {
    enumerable: true,
    get: function() {
        return Translate;
    }
});
var _lodash = require("lodash");
var _every = /*#__PURE__*/ _interop_require_default(require("lodash/every"));
var _isPlainObject = /*#__PURE__*/ _interop_require_default(require("lodash/isPlainObject"));
var _isString = /*#__PURE__*/ _interop_require_default(require("lodash/isString"));
var _isUndefined = /*#__PURE__*/ _interop_require_default(require("lodash/isUndefined"));
var _keys = /*#__PURE__*/ _interop_require_default(require("lodash/keys"));
var _pickBy = /*#__PURE__*/ _interop_require_default(require("lodash/pickBy"));
var _values = /*#__PURE__*/ _interop_require_default(require("lodash/values"));
var _base = require("./base");
var _transform = require("./transform");
var _cssvalue = require("../utils/css-value");
var _dimension = require("../utils/dimension");
var _generateguard = require("../utils/generate-guard");
function _assert_this_initialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _get_prototype_of(o) {
    _get_prototype_of = Object.setPrototypeOf ? Object.getPrototypeOf : function getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _get_prototype_of(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _set_prototype_of(subClass, superClass);
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function _possible_constructor_return(self, call) {
    if (call && (_type_of(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assert_this_initialized(self);
}
function _set_prototype_of(o, p) {
    _set_prototype_of = Object.setPrototypeOf || function setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _set_prototype_of(o, p);
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _is_native_reflect_construct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;
    try {
        Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
        return true;
    } catch (e) {
        return false;
    }
}
function _create_super(Derived) {
    var hasNativeReflectConstruct = _is_native_reflect_construct();
    return function _createSuperInternal() {
        var Super = _get_prototype_of(Derived), result;
        if (hasNativeReflectConstruct) {
            var NewTarget = _get_prototype_of(this).constructor;
            result = Reflect.construct(Super, arguments, NewTarget);
        } else {
            result = Super.apply(this, arguments);
        }
        return _possible_constructor_return(this, result);
    };
}
var Translate = /*#__PURE__*/ function(Base) {
    "use strict";
    _inherits(Translate, Base);
    var _super = _create_super(Translate);
    function Translate() {
        _class_call_check(this, Translate);
        var _this;
        _this = _super.apply(this, arguments);
        _define_property(_assert_this_initialized(_this), "isProcessableValue", (0, _generateguard.generateGuard)(_isString.default, _isUndefined.default));
        _define_property(_assert_this_initialized(_this), "isProcessableValues", (0, _generateguard.generateGuard)([
            _isPlainObject.default,
            function(maybe) {
                return (0, _every.default)((0, _keys.default)(maybe), _isString.default);
            },
            function(maybe) {
                return (0, _every.default)((0, _values.default)(maybe), _this.isProcessableValue);
            }
        ]));
        _define_property(_assert_this_initialized(_this), "normaliseValues", function(values) {
            return _this.isProcessableValues(values) ? (0, _lodash.chain)(values).mapValues(function(duration) {
                return (0, _cssvalue.normaliseLengthPercentageValue)(duration);
            }).pickBy(function(duration, modifier) {
                return (0, _isString.default)(modifier) && modifier !== "" && (0, _isString.default)(duration);
            }).value() : {};
        });
        _define_property(_assert_this_initialized(_this), "utilities", function() {
            var values = _this.normaliseValues(_this.api.theme("translate"));
            var cssDeclarations = _this.legacy ? Translate.legacyDeclarations() : Translate.declarations();
            _this.api.matchUtilities({
                "translate-x": function(value) {
                    return _object_spread({
                        "@defaults transform": {},
                        "--tw-translate-x": value
                    }, cssDeclarations);
                },
                "translate-y": function(value) {
                    return _object_spread({
                        "@defaults transform": {},
                        "--tw-translate-y": value
                    }, cssDeclarations);
                }
            }, {
                values: values,
                supportsNegativeValues: true
            });
            _this.api.matchUtilities({
                "translate-z": function(value) {
                    return _object_spread({
                        "@defaults transform": {},
                        "--tw-translate-z": value
                    }, cssDeclarations);
                }
            }, {
                values: (0, _pickBy.default)(values, function(value) {
                    return !value.endsWith("%");
                }),
                supportsNegativeValues: true
            });
        });
        return _this;
    }
    return Translate;
}(_base.Base);
_define_property(Translate, "defaultFunctionValues", {
    translateX: "var(--tw-translate-x)",
    translateY: "var(--tw-translate-y)",
    translateZ: "var(--tw-translate-z)"
});
_define_property(Translate, "normaliseFunctionValues", function() {
    var _ref = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {}, dimension = _ref.dimension, translateX = _ref.translateX, translateY = _ref.translateY, translateZ = _ref.translateZ;
    var safeValues = [
        (0, _cssvalue.normaliseLengthPercentageValue)(translateX, Translate.defaultFunctionValues.translateX),
        (0, _cssvalue.normaliseLengthPercentageValue)(translateY, Translate.defaultFunctionValues.translateY)
    ];
    if ((0, _dimension.normaliseDimension)(dimension) === "3d") {
        safeValues.push((0, _cssvalue.normaliseLengthValue)(translateZ, Translate.defaultFunctionValues.translateZ));
    }
    return safeValues.join(" ");
});
_define_property(Translate, "declarations", function() {
    var values = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {};
    return {
        translate: Translate.normaliseFunctionValues(values)
    };
});
_define_property(Translate, "legacyDeclarations", function() {
    return _transform.Transform.legacyDeclarations();
});
